torchmetrics/text/squad.py

```diff
@@ -0,0 +1,181 @@
+
+                Example target:
+
+                .. code-block:: python
+
+                    {
+                        'answers': [{'answer_start': [1], 'text': ['This is a test answer']}],
+                        'id': '1',
+                    }
+
+                Reference SQuAD Format:
+
+                .. code-block:: python
+
+                    {
+                        'answers': {'answer_start': [1], 'text': ['This is a test text']},
+                        'context': 'This is a test context.',
+                        'id': '1',
+                        'question': 'Is this a test?',
+                        'title': 'train test'
+                    }
+
+        Raises:
+            KeyError:
+                If the required keys are missing in either predictions or targets.
+        """
+        if isinstance(preds, Dict):
+            preds = [preds]
+
+        if isinstance(targets, Dict):
+            targets = [targets]
+
+        for pred in preds:
+            keys = pred.keys()
+            if "prediction_text" not in keys or "id" not in keys:
+                raise KeyError(
+                    "Expected keys in a single prediction are 'prediction_text' and 'id'."
+                    "Please make sure that 'prediction_text' maps to the answer string and 'id' maps to the key string."
+                )
+
+        for target in targets:
+            keys = target.keys()
+            if "answers" not in keys or "id" not in keys:
+                raise KeyError(
+                    "Expected keys in a single target are 'answers' and 'id'."
+                    "Please make sure that 'answers' maps to a `SQuAD` format dictionary and 'id' maps to the key"
+                    "string.\n"
+                    "SQuAD Format: "
+                    f"{SQuAD_FORMAT}"
+                )
+
+            answers: Dict[str, Any] = target["answers"]  # type: ignore
+            if "text" not in answers.keys():
+                raise KeyError(
+                    "Expected keys in a 'answers' are 'text'."
+                    "Please make sure that 'answer' maps to a `SQuAD` format dictionary.\n"
+                    "SQuAD Format: "
+                    f"{SQuAD_FORMAT}"
+                )
+
+        preds_dict = {prediction["id"]: prediction["prediction_text"] for prediction in preds}
+        targets_dict = [
+            {
+                "paragraphs": [
+                    {
+                        "qas": [
+                            {
+                                "answers": [
+                                    {"text": answer_text} for answer_text in target["answers"]["text"]  # type: ignore
+                                ],
+                                "id": target["id"],
+                            }
+                            for target in targets
+                        ]
+                    }
+                ]
+            }
+        ]
```

Comment:
This is duplicate code between functional and modular. Please either move this into the `_squad_update` function or make a separate `_squad_input_check` function.